### ISSUE TITLE
Federation: Google Cloud DNS provider does not handle missing domains correctly

### ISSUE THREAD

**Issue by quinton-hoole** [kubernetes/kubernetes#51139]
/kind bug

**What happened**:
walteraa [6:55 AM] 
Hi, The controller manager of the federation using Google cloud DNS is crashing, the log showed is that:

```I0822 13:17:33.997888       1 clouddns.go:86] Using DefaultTokenSource &oauth2.reuseTokenSource{new:jwt.jwtSource{ctx:(*context.emptyCtx)(0xc42001c550), conf:(*jwt.Config)(0xc420288e00)}, mu:sync.Mutex{state:0, sema:0x0}, t:(*oauth2.Token)(nil)}
I0822 13:17:33.998181       1 clouddns.go:100] Successfully got DNS service: &{0xc420e19b00 https://www.googleapis.com/dns/v1/projects/  0xc42000e840 0xc42000e848 0xc42000e850 0xc42000e858}
E0822 13:17:35.621941       1 dns.go:97] Failed to retrieve DNS zone: error querying for DNS zones: googleapi: got HTTP response code 404 with body: Not Found
F0822 13:17:35.622009       1 controllermanager.go:144] Failed to start service dns controller: error querying for DNS zones: googleapi: got HTTP response code 404 with body: Not Found
```

the hyperkube command is that:
```/hyperkube
      federation-controller-manager
      --dns-provider=google-clouddns
      --federation-name=fed-keepalived
      --kubeconfig=/etc/federation/controller-manager/kubeconfig
      --master=https://fed-keepalived-apiserver
      --zone-name=kfed.walteralves.me.
```

I have created this `kfed.walteralves.me.` in the Google DNS

Someone knows why the DNS zone query is getting the 404 error?? :disappointed:

**What you expected to happen**:

It should have created the missing domain.

**How to reproduce it (as minimally and precisely as possible)**:

As per above.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): 1.7.4
- Cloud provider or hardware configuration**: OpenStack (irrelevant)
- OS (e.g. from /etc/os-release): irrelevant
- Kernel (e.g. `uname -a`): irrelevant
- Install tools: irrelevant
- Others: irrelevant

**Comment by quinton-hoole** [comment]
The bug is related to this:

https://github.com/kubernetes/kubernetes/blob/master/federation/pkg/federation-controller/service/dns/dns.go#L227

The above code assumes that the provider code returns no error but an empty list if the domain does not exist, but instead the provider returns a 404.  Need to fix the provider and/or the above code.

**Comment by quinton-hoole** [comment]
I am unlikely to have time to work on this soon, so if anyone else wants to, go right ahead.

**Comment by walteraa** [comment]
An additional information: I have created a `secret` which is based on the credential JSON file created through [this](https://developers.google.com/identity/protocols/application-default-credentials) instruction. After that, I have configured the Federation controller manager to mount this secret as a volume in the container and set the environment variable `GOOGLE_APPLICATION_CREDENTIALS` to the JSON file inside this volume.

Before doing that, I was getting an authentication error so the google DNS provider couldn't be instantiated, and after doing the instructions that I mentioned before, this authentication error was solved but now I'm getting 404 error when Google DNS provider tries access the specified managed zone.
